core/crypto/abac/attributes.go

```diff
@@ -0,0 +1,271 @@
+	if err != nil {
+		return nil, encrypted, err
+	}
+	return value, encrypted, nil
+}
+
+//EncryptAttributeValue encrypts "attributeValue" using "attributeKey"
+func EncryptAttributeValue(attributeKey []byte, attributeValue []byte) ([]byte, error) {
+	value := append(attributeValue, padding...)
+	return primitives.CBCPKCS7Encrypt(attributeKey, value)
+}
+
+//getAttributeKey returns the attributeKey derived from the preK0 to the attributeName.
+func getAttributeKey(preK0 []byte, attributeName string) []byte {
+	return primitives.HMACTruncated(preK0, []byte(attributeName), 32)
+}
+
+//EncryptAttributeValuePK0 encrypts "attributeValue" using a key derived from preK0.
+func EncryptAttributeValuePK0(preK0 []byte, attributeName string, attributeValue []byte) ([]byte, error) {
+	attributeKey := getAttributeKey(preK0, attributeName)
+	return EncryptAttributeValue(attributeKey, attributeValue)
+}
+
+//DecryptAttributeValue decrypts "encryptedValue" using "attributeKey" and return the decrypted value.
+func DecryptAttributeValue(attributeKey []byte, encryptedValue []byte) ([]byte, error) {
+	value, err := primitives.CBCPKCS7Decrypt(attributeKey, encryptedValue)
+	if err != nil {
+		return nil, err
+	}
+	lenPadding := len(padding)
+	lenValue := len(value)
+	if lenValue < lenPadding {
+		return nil, errors.New("Error invalid value. Decryption verification failed.")
+	}
+	lenWithoutPadding := lenValue - lenPadding
+	if bytes.Compare(padding[0:lenPadding], value[lenWithoutPadding:lenValue]) != 0 {
+		return nil, errors.New("Error generating decryption key for value. Decryption verification failed.")
+	}
+	value = value[0:lenWithoutPadding]
+	return value, nil
+}
+
+//getKAndValueForAttribute derives K for the attribute "attributeName", checks the value padding and returns both key and decrypted value
+func getKAndValueForAttribute(attributeName string, preK0 []byte, cert *x509.Certificate) ([]byte, []byte, error) {
+	headerKey := getAttributeKey(preK0, HeaderAttributeName)
+	value, encrypted, err := ReadTCertAttribute(cert, attributeName, headerKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	attributeKey := getAttributeKey(preK0, attributeName)
+	if encrypted {
+		value, err = DecryptAttributeValue(attributeKey, value)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	return attributeKey, value, nil
+}
+
+//GetKForAttribute derives the K for the attribute "attributeName" and returns the key
+func GetKForAttribute(attributeName string, preK0 []byte, cert *x509.Certificate) ([]byte, error) {
+	key, _, err := getKAndValueForAttribute(attributeName, preK0, cert)
+	return key, err
+}
+
+//GetValueForAttribute derives the K for the attribute "attributeName" and returns the value
+func GetValueForAttribute(attributeName string, preK0 []byte, cert *x509.Certificate) ([]byte, error) {
+	_, value, err := getKAndValueForAttribute(attributeName, preK0, cert)
+	return value, err
+}
+
+func createABACHeaderEntry(preK0 []byte) (*pb.ABACMetadataEntry, error) {
```

Comment:
My comment is for the `createABACHeaderEntry` function. It never calls `getKAndValueForAttribute`